### ISSUE TITLE
"SELEC FROM" syntax error

### ISSUE THREAD

**Issue by tjwebb** [balderdashy/waterline-sequel#58]
https://github.com/balderdashy/waterline-sequel/blob/daa4646d3e28fed11f835adb4e7675a63a08cde7/sequel/select.js#L108

This line can cause invalid SQL to be generated. There are other reports of this happening 'round the internets, but I'm tracing it to the line above.

- http://stackoverflow.com/q/29067583/291180
- https://groups.google.com/forum/#!topic/sailsjs/D1u-R4cDwxA

**Comment by particlebanana** [comment]
Looks like it could happen when there are no attributes defined. 

This line https://github.com/balderdashy/waterline-sequel/blob/daa4646d3e28fed11f835adb4e7675a63a08cde7/sequel/select.js#L73 builds up an array of keys to select then loops through and appends to the query: https://github.com/balderdashy/waterline-sequel/blob/daa4646d3e28fed11f835adb4e7675a63a08cde7/sequel/select.js#L95-L105

So if it's empty it would remove extra characters. Not that you would get anything back from an empty query anyway but you wouldn't get the error message.

**Comment by acis** [comment]
I got this error as well. It happens when selecting n*m attributes. A workaround for this is to add another simple type field to the select. 

Owner.find({select: ['pets', 'addresses']})  throws the error,  but 
Owner.find({select: ['id', 'pets', 'addresses']}) works fine.

**Comment by devinivy** [comment]
Ah, that makes sense.  Seems like we could fix that and give a proper error.  But what does it mean to select a field that represents a collection?  You'll want to use `populate()` instead.

**Comment by acis** [comment]
I just don't need the other fields :)

**Comment by sailsbot** [comment]
Thanks for posting, @tjwebb.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
@tjwebb no longer an issue?

**Comment by devinivy** [comment]
@particlebanana this is still an issue.  It happens when you select no physical attributes– such as only selecting attributes that represent collections.

`error: syntax error at or near "SELEC"`

**Comment by dash-** [comment]
This issue is caused by a less-than-ideal approach to joining groups of phrases.  It's an attempt to remove a trailing comma, I believe.  Instead, we should just build an array of things to be joined in commas, and use the Array.prototype.join method to join them instead of building a string and then attempting to remove the comma.

**Comment by mikermcneil** [comment]
@devinivy @dash- I believe this is resolved in https://github.com/particlebanana/waterline-query-builder

We'll be using that in the next version of sails-mysql and sails-postgresql (cc @particlebanana)

**Comment by dash-** [comment]
Cool.

I wish I'd have realized that prior to making waterline-sequel-derby based
on waterline-sequel, and prior to making sails-derby based on the former.
If waterline-query-builder is more flexible / configurable than its
predecessor, I may just be able to move sails-derby over to
waterline-query-builder.

The key issue I had with waterline-sequel is that Apache Derby uses single
quotes for literals, and the literal quote character wasn't configurable.
But I'll check out waterline-query-builder and see if I can maybe just drop
waterline-sequel-derby altogether.

**Comment by particlebanana** [comment]
@dash- The new library uses Knex so it's only available for databases supported by that.

**Comment by mikermcneil** [comment]
@dash-  Great work on sails-derby!  Sorry for the confusion.  Like @particlebanana mentioned, the new SQL builder is only for databases supported by knex; and we're only planning on switching over sails-mysql and sails-postgresql in the immediate term.  We'll also still be building some queries (e.g. DDL) without the help of the new sql builder, at least for now.

**Comment by davemackintosh** [comment]
Is there a fix or workaround for this, it's affecting a few projects I'm working on as well as Multicolour :/
